src/lib/exports.ts

```diff
@@ -403,3 +426,350 @@ export const exportNoteAsPdfFile = async (
+          noteDoc as NoteDoc,
+          includeFrontMatter
+        )
+        const mdNoteLocation = join(
+          join(join(rootDir, noteExportFolder), mdName)
+        )
+        notifyExportOperationState(
+          prepareExportOperationState(
+            noteDoc.title,
+            rootDir,
+            mdNoteLocation,
+            currentNote,
+            exportingNoteCount
+          )
+        )
+        await writeNoteData(mdString, mdNoteLocation, exportErrors)
+        break
+
+      case 'pdf':
+      default:
+        const pdfName = `${filenamify(getValidNoteTitle(noteDoc))}.pdf`
+        const pdfNoteLocation = join(
+          join(join(rootDir, noteExportFolder), pdfName)
+        )
+        const pdfBuffer = await convertNoteDocToPdfBuffer(
+          noteDoc as NoteDoc,
+          preferences,
+          pushMessage,
+          (src) => getAttachmentData(storage, src),
+          previewStyle
+        )
+
+        notifyExportOperationState(
+          prepareExportOperationState(
+            noteDoc.title,
+            rootDir,
+            pdfNoteLocation,
+            currentNote,
+            exportingNoteCount
+          )
+        )
+        await writeNoteData(pdfBuffer, pdfNoteLocation, exportErrors)
+        break
+    }
+  }
+}
+
+export interface ExportOperationState {
+  exportOperation: string
+  exportSubOperationTitle: string
+  exportSubOperationContent: string
+  exportProgressValue: number
+}
+
+function prepareExportOperationState(
+  noteTitle: string,
+  rootDir: string,
+  noteLocation: string,
+  currentNote: number,
+  totalNotesForExport: number
+) {
+  return {
+    exportSubOperationTitle: `Exporting note: ${
+      noteTitle ? noteTitle : 'No title'
+    } (${currentNote}/${totalNotesForExport})`,
+    exportSubOperationContent: `Location: ${noteLocation.replace(
+      rootDir,
+      '.'
+    )}`,
+  }
+}
+
+async function createExportDirectories(
+  folderMap: ObjectMap<FolderDoc>,
+  rootDir: string,
+  folderName: string,
+  folderPathname: string,
+  recursiveExport: boolean,
+  isExportCanceled: () => boolean,
+  notifyActiveExportOperation: (
+    newExportOperationState: Partial<ExportOperationState>
+  ) => any,
+  storageExport = false
+) {
+  for (const folderPath of keys(folderMap)) {
+    if (isExportCanceled()) {
+      return
+    }
+
+    if (folderPath == '/' && !storageExport) {
+      continue
+    }
+
+    if (
+      !storageExport &&
+      ((recursiveExport && !folderPath.startsWith(folderPathname)) ||
+        (!recursiveExport && folderPath != folderPathname))
+    ) {
+      continue
+    }
+
+    const fullPath = join(
+      rootDir,
+      folderPath.substring(folderPath.indexOf(folderName))
+    )
+
+    notifyActiveExportOperation({
+      exportSubOperationTitle: `Creating directory: ${fullPath.replace(
+        rootDir,
+        '.'
+      )}`,
+    })
+    await mkdir(fullPath)
+  }
+}
+
+export async function batchExport(
```

Comment:
Too many arguments

Reply:
Will try to refactor it in smaller functions, but a lot of arguments are needed because it calls each export type, for each note, creates directories etc.

I think we could design communication between component and exports.ts in a more clear way, and this could reduce the arguments count.